File: Rust/15. 三数之和.rs
impl Solution {
    pub fn three_sum(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        nums.sort();
        let mut result = Vec::new();
        let mut first:usize = 0;
        while nums.len() > 2 && first < (nums.len() - 2){
            if first > 0 && nums[first]==nums[first-1]{
                first = first + 1;
            }
            else{
                let mut second:usize = first + 1;
                let mut third: usize = nums.len()-1;
                while second < third{
                    if nums[first] + nums[second] + nums[third] == 0{
                        result.push(vec![nums[first],nums[second],nums[third]]);
                        while second < third && nums[second+1] == nums[second]{
                            second = second + 1;
                        }
                        while second < third && nums[third-1] == nums[third]{
                            third = third - 1;
                        }
                        second = second + 1;
                        third = third - 1;
                    } else if nums[first] + nums[second] + nums[third] < 0{
                        second = second+1;
                    }else{
                        third = third-1;
                    }
                }
                first = first + 1;
            }
        }
        result
    }
}

File: CPlusPlus/15. 三数之和.cpp
// 解法 1： 使用 set 去重
#include <vector>
#include <set>
#include <algorithm>

using namespace std;
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        if(nums.size()<3)
            return vector<vector<int>>();
        std::sort(std::begin(nums),std::end(nums));
        for(auto i{std::begin(nums)}; i!=std::end(nums)-2;++i){
            auto j{i+1};
            auto k{std::end(nums)-1};
            while(j < k ){
                if((*i)+ (*j)+(*k) == 0){
                    res.insert(std::vector<int>{*i,*j,*k});
                    ++j;
                    --k;
                }
                else if((*i)+ (*j)+(*k) > 0){
                    --k;
                }
                else ++j;
            }
        }
        return vector<vector<int>>(res.begin(),res.end());
    }
private:
    set<vector<int>> res;
};
// 解法 2： 挪动指针的时候判断和前者是否相等， 相等则跳过
class BetterSolution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res{};
        int n = nums.size();
        for(int i = 0; i < n-2; ++i){
            if(i > 0 && nums[i] == nums[i-1]) continue;
            int left = i + 1;
            int right = n -1;
            while(left < right){
                if(left > i+1 && nums[left] == nums[left-1]){
                    ++left;
                }else if(right < nums.size()-1 && nums[right] == nums[right+1]){
                    --right;
                }else if(nums[left]+nums[right]+nums[i] > 0){
                    --right;
                }else if(nums[left]+nums[right]+ nums[i] == 0){
                    res.emplace_back(vector<int>{nums[i],nums[left],nums[right]});
                    ++left;
                    --right;
                }else{
                    ++left;
                }
            }
        }
        return res;
    }
};
